lib/pharos/phases/configure_master.rb

```diff
@@ -1,361 +1,69 @@
 # frozen_string_literal: true
 
-require "base64"
-
 module Pharos
   module Phases
     class ConfigureMaster < Pharos::Phase
       title "Configure master"
 
-      KUBE_DIR = '/etc/kubernetes'
-      PHAROS_DIR = '/etc/pharos'
-      SHARED_CERT_FILES = %w(ca.crt ca.key sa.key sa.pub).freeze
-      AUTHENTICATION_TOKEN_WEBHOOK_CONFIG_DIR = '/etc/kubernetes/authentication'
-      AUDIT_CFG_DIR = (PHAROS_DIR + '/audit').freeze
-      SECRETS_CFG_DIR = (PHAROS_DIR + '/secrets-encryption').freeze
-      SECRETS_CFG_FILE = (SECRETS_CFG_DIR + '/config.yml').freeze
-
-      def call
-        logger.info { "Checking if Kubernetes control plane is already initialized ..." }
-        if install?
-          logger.info { "Kubernetes control plane is not initialized, proceeding to initialize ..." }
-          install
-        elsif upgrade?
-          logger.info { "Upgrading Kubernetes control plane ..." }
-          upgrade
-        else
-          logger.info { "Kubernetes control plane is up-to-date." }
-          configure
-        end
-      end
-
-      def install?
-        !@ssh.file("/etc/kubernetes/admin.conf").exist?
+      def kubeadm_config
+        @kubeadm_config ||= Pharos::Kubeadm::Config.new(@config, @host)
       end
 
-      def upgrade?
-        manifest = File.join(KUBE_DIR, 'manifests', 'kube-apiserver.yaml')
-        file = @ssh.file(manifest)
-        return false unless file.exist?
-        return false if file.read.match?(/kube-apiserver-.+:v#{Pharos::KUBE_VERSION}/)
-
-        true
-      end
-
-      def install
-        configure_kubelet
-
-        cfg = generate_config
-
-        # Copy etcd certs over if needed
-        if @config.etcd&.certificate
-          logger.info { "Pushing external etcd certificates ..." }
-          copy_external_etcd_certs
-        end
+      def call
+        push_external_etcd_certs if @config.etcd&.certificate
 
         push_audit_config if @config.audit&.server
 
-        # Generate and upload authentication token webhook config file if needed
         push_authentication_token_webhook_config if @config.authentication&.token_webhook
-
-        copy_kube_certs
-
-        logger.info { "Initializing control plane ..." }
-
-        @ssh.tempfile(content: cfg.to_yaml, prefix: "kubeadm.cfg") do |tmp_file|
-          @ssh.exec!("sudo kubeadm init --ignore-preflight-errors all --skip-token-print --config #{tmp_file}")
-        end
-
-        cache_kube_certs
```

Comment:
Bug in master: the `cache_kube_certs`/`copy_kube_certs` is not handled correctly if you interrupt the `pharos-cluster up` during the initial install after the first master node was installed... when you later retry, the first master host will already be `installed?`, so it will skip the `install` => `cache_kube_certs`, and the remaining master hosts will use different certs than the first master node.